File: 2018/02/solution.py
#!/usr/bin/env python3
"""AoC Day 2"""

import sys

class Counter(dict):
    """Counter"""
    def __missing__(self, key):
        return 0

def load(argc, argv):
    if argc != 2:
        print('usage: {} input-file'.format(argv[0]))
        exit(0)

    with open(argv[1], 'r') as fdd:
        dat = fdd.read().splitlines()

    return dat

def part1(argc, argv):
    """ Part 1 entry point"""

    dat = load(argc, argv)

    two_cnt = 0
    three_cnt =0
    for word in dat:
        cnt = Counter()
        for c in word:
            cnt[c] += 1

        got_two = 0
        got_three = 0
        for c in cnt:
            if cnt[c] == 2 and got_two == 0:
                two_cnt += 1
                got_two = 1
            elif cnt[c] == 3 and got_three == 0:
                three_cnt += 1
                got_three = 1


    print('Part 1: {}'.format(three_cnt * two_cnt))

def part2(argc, argv):
    """ Part 2 entry point"""
    dat = load(argc, argv)

    num_words = len(dat)
    for i in range(num_words):
        for j in range(i + 1, num_words):
            diff = 0
            for (iw, jw) in zip(dat[i], dat[j]):
                if iw != jw:
                    diff += 1

            if diff == 1:
                print('{} : {}'.format(dat[i], dat[j]))


if __name__ == "__main__":
    part1(len(sys.argv), sys.argv)
    part2(len(sys.argv), sys.argv)

File: 2018/02/solution.rb
#!/usr/bin/env ruby

# Read in and convert to integers
dat = File.readlines('input.txt').map(&:strip)

# Solve Part 1
two_cnt = 0
three_cnt = 0
dat.each do |word|
    dict = { }
    word.each_char do |ch| 
        if dict[ch].nil?
            dict[ch] = 1
        else
            dict[ch] += 1
        end
    end

    if dict.has_value?(2)
        two_cnt += 1
    end

    if dict.has_value?(3)
        three_cnt += 1
    end
end

part1 = two_cnt * three_cnt
puts "Part 1: #{part1}"

# Solve Part 2

for i in 0..(dat.count - 1) do
    for j in (i + 1)..(dat.count - 1) do
        diff = 0
        dat[i].chars.zip(dat[j].chars).each { |ic, jc| diff += 1 if ic != jc }

        if diff == 1
            puts "Part 2: #{dat[i]} : #{dat[j]}"
        end
    end
end
